Editorial review: Document WebSocketStream

chrisdavidmills opened the pull request:
<!-- 🙌 Thanks for contributing to MDN Web Docs. Adding details below will help us to merge your PR faster. -->

### Description

Chrome 124 supports the [`WebSocketStream` API](https://developer.chrome.com/docs/capabilities/web-apis/websocketstream) (see the associated [ChromeStatus entry](https://chromestatus.com/feature/5189728691290112)).

Note that `WebSocketStream` is currently supported in Chromium-based browsers, and is not yet part of a standard so I will mark it as non-standard for now. See https://github.com/whatwg/websockets/pull/48 for progress on that.

<!-- ✍️ Summarize your changes in one or two sentences -->

### Motivation

<!-- ❓ Why are you making these changes and how do they help readers? -->

### Additional details

<!-- 🔗 Link to release notes, vendor docs, bug trackers, source control, or other places providing more context -->

### Related issues and pull requests

<!-- 🔨 If this fully resolves a GitHub issue, use "Fixes #123" -->
<!-- 👉 Highlight related pull requests using "Relates to #123" -->
<!-- ❗ If another pull request should be merged first, use "**Depends on:** #123" -->


<!-- 👷‍♀️ After submitting, go to the "Checks" tab of your PR for the build status -->

chrisdavidmills committed:
Document WebSocketStream

chrisdavidmills committed:
Add WebSocket interface documentation

chrisdavidmills committed:
Merge branch 'main' into websocketstream

tomayac reviewed (commented):
This is in a really good shape. Just some nits. 

Maybe ping @ricea for a spec editor review.

chrisdavidmills committed:
Update files/en-us/web/api/websockets_api/using_websocketstream/index.md

Co-authored-by: Thomas Steiner <tomac@google.com>

chrisdavidmills committed:
Update files/en-us/web/api/websockets_api/using_websocketstream/index.md

Co-authored-by: Thomas Steiner <tomac@google.com>

chrisdavidmills committed:
Update files/en-us/web/api/websockets_api/using_websocketstream/index.md

Co-authored-by: Thomas Steiner <tomac@google.com>

chrisdavidmills committed:
Update files/en-us/web/api/websockets_api/using_websocketstream/index.md

Co-authored-by: Thomas Steiner <tomac@google.com>

chrisdavidmills committed:
Update files/en-us/web/api/websockets_api/using_websocketstream/index.md

Co-authored-by: Thomas Steiner <tomac@google.com>

chrisdavidmills committed:
Update files/en-us/web/api/websockets_api/using_websocketstream/index.md

Co-authored-by: Thomas Steiner <tomac@google.com>

chrisdavidmills committed:
Update files/en-us/web/api/websocketstream/closed/index.md

Co-authored-by: Thomas Steiner <tomac@google.com>

chrisdavidmills committed:
Update files/en-us/web/api/websocketstream/opened/index.md

Co-authored-by: Thomas Steiner <tomac@google.com>

chrisdavidmills committed:
Update files/en-us/web/api/websocketstream/url/index.md

Co-authored-by: Thomas Steiner <tomac@google.com>

chrisdavidmills committed:
Update files/en-us/web/api/websocketstream/websocketstream/index.md

Co-authored-by: Thomas Steiner <tomac@google.com>

chrisdavidmills committed:
make protocol usage consistent ascross example snippets

chrisdavidmills commented:
Thanks, @tomayac! I accepted all of your suggestions. I also checked and updated a few example snippets to use `wss://` and `self` for feature detection to keep things consistent.

Also, because we need to specify `ws://` in the code example for it to work as written, I've added a note into the `WebSocketStream` guide to draw the reader's attention to this, just in case they use `wss://` and get stuck as a result.

I'm going to put this forward for editorial review now, but @ricea, feel free to have a look and make comments if you wish. It should take a little while for someone to pick it up, so there is time.

wbamberg commented on files/en-us/web/api/websockets_api/using_websocketstream/index.md:
````diff
@@ -0,0 +1,273 @@
+
+```js
+setTimeout(async () => {
+  wss.close({
+    code: 1000,
+    reason: "That's all folks",
````
We say a few times that it's preferable to use `abort()`, maybe better to show that?

chrisdavidmills replied:
We also say that close() is used when providing a custom code and/or reason. In this case we are providing a custom reason, so I think this is OK.

wbamberg replied:
To be pedantic, which I always am, it says "custom code **and** reason".

chrisdavidmills replied:
OK, all relevant instances now changed to "and/or".

wbamberg commented on files/en-us/web/api/websockets_api/using_websocketstream/index.md:
```diff
@@ -0,0 +1,273 @@
+    <h2>WebSocketStream Test</h2>
+    <p>Sends a ping every five seconds</p>
+    <div id="output"></div>
+    <script>
```
I think it would be easier to read this if the JS were in an external script.

chrisdavidmills replied:
Possibly so, but this would also make it slightly more annoying for a user to copy and paste and chuck into a file for testing purposes. I think it is better to leave as-is, by this token.

wbamberg commented on files/en-us/web/api/websockets_api/using_websocketstream/index.md:
```diff
@@ -0,0 +1,273 @@
+slug: Web/API/WebSockets_API/Using_WebSocketStream
+page-type: guide
+status:
+  - non-standard
```
also experimental?

chrisdavidmills replied:
OK, makes sense. I've also added it to the `WebSocketStream` interface pages.

wbamberg replied:
FWIW this should get added automatically from the BCD. But only in reference pages, not in guide pages.

chrisdavidmills replied:
Ah, OK. I keep forgetting what is automatically added and what isn't.

wbamberg commented on files/en-us/web/api/websockets_api/using_websocketstream/index.md:
```diff
@@ -0,0 +1,273 @@
+
+The main bulk of our code is contained within the `start()` function, which we define and then immediately invoke. We await the {{domxref("WebSocketStream.opened", "opened")}} promise, then after it fulfills write a message to let the reader know the connection is successful and create {{domxref("ReadableStreamDefaultReader")}} and {{domxref("WritableStreamDefaultWriter")}} instances from the ret ...
+
+Next, we `write()` a `"ping"` value to the socket and communicate that to the user. At this point, the server will respond with a `"pong"` message. We await the `read()` of the read message, communicate it to the user, then write another `"ping"` to the server after a timeout of 5 seconds. This continues the `"ping"`/`"pong"` loop indefinitely.
```
```suggestion
Next, we `write()` a `"ping"` value to the socket and communicate that to the user. At this point, the server will respond with a `"pong"` message. We await the `read()` of the reasponse, communicate it to the user, then write another `"ping"` to the server after a timeout of 5 seconds. This continues the `"ping"`/`"pong"` loop indefinitely.
```

wbamberg commented on files/en-us/web/api/websockets_api/using_websocketstream/index.md:
```diff
@@ -0,0 +1,273 @@
+> [!NOTE]
+> To get the example working, you'll also need a server component. We wrote our client to work along with the Deno server explained in [Writing a WebSocket server in JavaScript (Deno)](/en-US/docs/Web/API/WebSockets_API/Writing_a_WebSocket_server_in_JavaScript_Deno), but any compatible server will do.
+
+First we grab a reference to a DOM element into which we will write output messages, and define a utility function that writes a message to it:
```
I'm a bit worried that this duplication of the code will make it hard to maintain. I understand though that just quoting the whole block then talking through it all might be indigestible, so idk what is for the best.

chrisdavidmills replied:
I appreciate that, but I've not got a better solution. Leaving as-is.

wbamberg reviewed (commented):
Here are comments on the non-reference pages.

wbamberg commented on files/en-us/web/api/websocketstream/closed/index.md:
````diff
@@ -0,0 +1,88 @@
+
+## Examples
+
+```js
````
I think it would be better to have a smaller more focused example here. Even just something like:

```js
  const url = "wss://127.0.0.1/";
  const socket = new WebSocketStream(url);

  socket.closed.then((result) => 
    console.log("Socket closed", result.closeCode, result.reason)
);
```

wbamberg commented on files/en-us/web/api/websockets_api/using_websocketstream/index.md:
```diff
@@ -0,0 +1,273 @@
+  writeToScreen(
+    `DISCONNECTED: code ${result.closeCode}, message "${result.reason}"`,
+  );
+  console.error("Socket closed", result.closeCode, result.reason);
```
Why not console.log? This isn't necessarily an error, is it?

chrisdavidmills replied:
True. I've updated all similar instances of `console.error` to `console.log` across these docs.

wbamberg commented on files/en-us/web/api/websocketstream/index.md:
```diff
@@ -0,0 +1,91 @@
+- {{domxref("WebSocketStream.url", "url")}} {{ReadOnlyInline}}
+  - : Returns the URL of the WebSocket server that the `WebSocketStream` instance was created with.
+- {{domxref("WebSocketStream.closed", "closed")}} {{ReadOnlyInline}}
+  - : Returns a {{jsxref("Promise")}} that fulfills with an object once the socket connection is closed. The object contains the closing code and reason sent by the server.
```
re "sent by the server", same comment as in the `closed` page, what if the client closes it?

chrisdavidmills replied:
Same answer as above. I'm not sure if this needs an update. I have modified it slightly, to "as sent by the server"

wbamberg commented on files/en-us/web/api/websocketstream/index.md:
```diff
@@ -0,0 +1,91 @@
+- {{domxref("WebSocketStream.closed", "closed")}} {{ReadOnlyInline}}
+  - : Returns a {{jsxref("Promise")}} that fulfills with an object once the socket connection is closed. The object contains the closing code and reason sent by the server.
+- {{domxref("WebSocketStream.opened", "opened")}} {{ReadOnlyInline}}
+  - : Returns a {{jsxref("Promise")}} that fulfills with an object once the socket connection is successfully opened. The object contains several useful features, including a {{domxref("ReadableStream")}} and a {{domxref("WritableStream")}} instance for receiving and sending data on the connection.
```
```suggestion
  - : Returns a {{jsxref("Promise")}} that fulfills with an object once the socket connection is successfully opened. Among other properties, the object includes a {{domxref("ReadableStream")}} and a {{domxref("WritableStream")}} instance for receiving and sending data on the connection.
```

Very picky. "several useful features" seems a bit odd. Should it be taken for granted that the properties are useful?

chrisdavidmills replied:
Yup, very picky. You are right too, which is even more annoying ;-)

I have updated the wording to "Among other features, this object contains a...", as the items listed aren't the actual property names, but the values they contain.

wbamberg commented on files/en-us/web/api/websocketstream/opened/index.md:
```diff
@@ -0,0 +1,87 @@
+{{APIRef("WebSockets API")}}{{non-standard_header}}
+
+The **`opened`** read-only property of the
+{{domxref("WebSocketStream")}} interface returns a {{jsxref("Promise")}} that fulfills with an object once the socket connection is successfully opened. The object contains several useful features, including a {{domxref("ReadableStream")}} and a {{domxref("WritableStream")}} instance for receiving and sending data on the connection.
```
"several useful features" again.

chrisdavidmills replied:
Same change made here.

wbamberg commented on files/en-us/web/api/websocketstream/opened/index.md:
```diff
@@ -0,0 +1,87 @@
+- `extensions`
+  - : A string representing any extensions applied to the `WebSocketStream`. Such extensions are not currently defined, but may be in the future. Currently returns an empty string.
+- `protocol`
+  - : A string representing any custom protocol used to open the current WebSocket connection, as specified in the [`protocols`](/en-US/docs/Web/API/WebSocketStream/WebSocketStream#protocols) option of the `WebSocketStream()` constructor. Returns an empty string if no custom protocol was used.
```
I guess this must be one of the ones given in the constructor?

chrisdavidmills replied:
I don't really understand the question. The text clearly says "as specified in the [`protocols`](/en-US/docs/Web/API/WebSocketStream/WebSocketStream#protocols) option of the `WebSocketStream()` constructor".

I've updated the end of the description to "if no custom protocol was set during instantiation", in case it is useful.

wbamberg replied:
I don't think "as specified in..." is very clear. `protocols` is an array, this is a single item. I'm assuming what happens is that the client passes, in `protocols`, an array of custom protocols it is prepared to work with, and the server responds with a single protocol from the array. But the docs don't say this, so I have to guess.

chrisdavidmills replied:
Ah, OK, I getcha now. I've done a bit of reading around this, and have updated this description to:

> A string representing the sub-protocol used to open the current WebSocket connection (chosen from the options specified in the [`protocols`](/en-US/docs/Web/API/WebSocketStream/WebSocketStream#protocols) option of the `WebSocketStream()` constructor). Returns an empty string if no sub-protocol has been used to open the connection (i.e. no sub-protocol options were included in the constructor call).

wbamberg commented on files/en-us/web/api/websocketstream/opened/index.md:
````diff
@@ -0,0 +1,87 @@
+## Examples
+
+```js
+const output = document.querySelector("#output");
````
As for closed, I think a smaller more focused example would be better.

chrisdavidmills replied:
I've cut it down quite a bit, but not as much as the `closed` example.

wbamberg reviewed (commented):
Comments for the other pages.

chrisdavidmills committed:
Merge branch 'main' into websocketstream

chrisdavidmills commented on files/en-us/web/api/websockets_api/index.md:
```diff
@@ -12,10 +14,16 @@ The **WebSocket API** is an advanced technology that makes it possible to open a
 > [!NOTE]
 > While a WebSocket connection is functionally somewhat similar to standard Unix-style sockets, they are not related.
 
+### WebSocketStream
+
+The {{domxref("WebSocketStream")}} API is a modern reimagining of WebSocket client-side JavaScript functionality. It is promise-based, and therefore simpler to work with than the traditional {{domxref("WebSocket")}} API in the modern JavaScript scosystem. In addition, it uses the [Streams API](/en-US/docs/Web/API/Streams_API) to handle receiving and sending messages, meaning that socket connectio ...
```
Good call. In my next commit, I have used your suggested structure as a basis, and filled in the missing details.

chrisdavidmills commented on files/en-us/web/api/websockets_api/using_websocketstream/index.md:
```diff
@@ -0,0 +1,273 @@
+
+{{DefaultAPISidebar("WebSockets API")}}{{non-standard_header}}
+
+The {{domxref("WebSocketStream")}} API is a modern reimagining of WebSocket client-side JavaScript functionality. It is promise-based, and therefore simpler to work with than the traditional {{domxref("WebSocket")}} API in the modern JavaScript scosystem. In addition, it uses the [Streams API](/en-US/docs/Web/API/Streams_API) to handle receiving and sending messages, meaning that socket connectio ...
```
The overview page no longer contains this paragraph. I have kept it on this page, but updated it to make it more concise and less marketing-y.

chrisdavidmills commented on files/en-us/web/api/websockets_api/using_websocketstream/index.md:
```diff
@@ -0,0 +1,273 @@
+
+The {{domxref("WebSocketStream")}} API is a modern reimagining of WebSocket client-side JavaScript functionality. It is promise-based, and therefore simpler to work with than the traditional {{domxref("WebSocket")}} API in the modern JavaScript scosystem. In addition, it uses the [Streams API](/en-US/docs/Web/API/Streams_API) to handle receiving and sending messages, meaning that socket connectio ...
+
+This article explains how to use the {{domxref("WebSocketStream")}} API to create a simple WebSocket client.
```
Argh, the curse of "simple"! Removed in next commit ;-)

chrisdavidmills commented on files/en-us/web/api/websockets_api/using_websocketstream/index.md:
```diff
@@ -0,0 +1,273 @@
+
+This article explains how to use the {{domxref("WebSocketStream")}} API to create a simple WebSocket client.
+
+> [!NOTE]
+> Backpressure is an issue with the traditional `WebSocket` API — it has no way to apply backpressure, therefore when messages arrive faster than the application can process them, the application will either fill up the device's memory by buffering those messages, become unresponsive due to 100% CPU usage, or both.
```
I have removed the note from here, and put a shortened version on the `WebSocket` page.

In terms of how to use it, it is just automatic for `WebSocketStream`. I do say "automatically" in the opening paragraph. Do you think I need to be a bit more explicit?

It is somewhat out of scope for this PR to start updating the Streams API docs. We could do that in a separate PR. Maybe file an issue for now?

wbamberg replied:
> In terms of how to use it, it is just automatic for WebSocketStream. I do say "automatically" in the opening paragraph. Do you think I need to be a bit more explicit?

Yes, I do. The doc links to https://developer.mozilla.org/en-US/docs/Web/API/Streams_API/Concepts#backpressure, which as I say is hard to understand, but does not make it obvious that the developer has to do nothing for backpressure to work.

I think it would be best to spell this out in the bit about "To read data from the socket, you can continuously call [ReadableStreamDefaultReader.read()](https://pr35548.content.dev.mdn.mozit.cloud/en-US/docs/Web/API/ReadableStreamDefaultReader/read) ...", and will leave a suggestion there.

chrisdavidmills replied:
OK, cool. I've also added a little qualifier here in parens - "(no additional action required by the developer)" - which will hopefully help too.

chrisdavidmills commented on files/en-us/web/api/websockets_api/using_websocketstream/index.md:
````diff
@@ -0,0 +1,273 @@
+To write data to the socket, you can use {{domxref("WritableStreamDefaultWriter.write()")}}:
+
+```js
+await writer.write("My message");
````
I was just including it because `write()` returns a promise. I guess I don't need it. Removed.

chrisdavidmills commented on files/en-us/web/api/websockets_api/using_websocketstream/index.md:
````diff
@@ -0,0 +1,273 @@
+await writer.write("My message");
+```
+
+To read data from the socket, you can continuously call {{domxref("ReadableStreamDefaultReader.read()")}} until the stream is done (i.e. until the returned `value` is `undefined`, and `done` is `true`):
````
I've used your text, as it makes more sense. 

The other part: when the stream is finished, `value` will equal `undefined`. But, this information isn't needed here. We are using `done` for the check in the code snippet.

chrisdavidmills commented on files/en-us/web/api/websockets_api/using_websocketstream/index.md:
````diff
@@ -0,0 +1,273 @@
+const { code, reason } = await wss.closed;
+```
+
+The promise rejects in the event of an unclean close operation.
````
It took me ages to find this detail, but I eventually found it in https://datatracker.ietf.org/doc/html/rfc6455#section-7.1.4.

I've just removed this line completely from this page, as it didn't seem that useful. I've added promise rejection information to the `opened` and `closed` property pages to describe the conditions under which the promise rejects. The `closed` description includes a description of what a clean close means, as it is relevant there.

chrisdavidmills commented on files/en-us/web/api/websockets_api/using_websocketstream/index.md:
````diff
@@ -0,0 +1,273 @@
+controller.abort();
+```
+
+Alternatively you can use the {{domxref("WebSocketStream.close()")}} method to close a connection, however this is mainly needed if you wish to specify a custom code and reason for the server to report:
````
updated

chrisdavidmills commented on files/en-us/web/api/websockets_api/using_websocketstream/index.md:
```diff
@@ -0,0 +1,273 @@
+const output = document.querySelector("#output");
+
+function writeToScreen(message) {
+  output.insertAdjacentHTML("beforeend", `<p>${message}</p>`);
```
It could be, yes, but this is a general security concern, not something specific to WebSockets. It would be the same if you somehow fetched a code snippet with something malicious in it? It is of course up to the developer to make sure they connect to a trusted server, have CSP or other mechanisms in place to prevent unwanted scripts running, etc.

I'm not convinced I need to include such details on this page. What do you think should be included?

wbamberg replied:
I was thinking, use `textContent` or something else safe, instead of `insertAdjacentHTML`.

> It is of course up to the developer to make sure they connect to a trusted server, have CSP or other mechanisms in place to prevent unwanted scripts running,

The issue about a trusted server is, how do you know you can trust it, and it hasn't been compromised? As for CSP, guidance I've read is that it is not a substitute for not removing XSS vectors.

It seems inherently unsafe to connect to a server expecting it to send you messages, then immediately (effectively) executing them without sanitizing them. And by putting this in an example we're telling devs that this is an OK thing to do.

wbamberg replied:
>  It would be the same if you somehow fetched a code snippet with something malicious in it? 

Yup, and I would make the same comment about a `fetch()` example that did this.

chrisdavidmills replied:
> use textContent or something else safe, instead of insertAdjacentHTML.

Right, this makes sense — I am not sure why I used `insertAdjacentHTML()`. I will update it to use textContent, which is what I normally use for such examples ;-\

chrisdavidmills commented on files/en-us/web/api/websockets_api/using_websocketstream/index.md:
````diff
@@ -0,0 +1,273 @@
+Next, we create an `if ... else` structure to feature detect `WebSocketStream` and output an informative message on non-supporting browsers:
+
+```js
+if ("WebSocketStream" in self) {
+  // supporting code path
+} else {
+  writeToScreen("Your browser does not support WebSocketStream");
+}
````
makes sense. I've updated the example to suit

chrisdavidmills commented on files/en-us/web/api/websockets_api/using_websocketstream/index.md:
```diff
@@ -0,0 +1,273 @@
+  const reader = readable.getReader();
+  const writer = writable.getWriter();
+
+  await writer.write("ping");
```
It's promise-based, so it won't work without it?

wbamberg replied:
You don't have to await to make it work: await just ensures that the promise fulfills before it executes the next line.

I mean, usually I see `await` being used if you want to make your code behave like synchronous code, for instance if it's returning something you need in the next line:

```js
const response = await fetch(url);
const data = await response.json();
```

(or as we are doing with `read()`, where we need the read value in the next line).

But in this case we're not, so what's functionally the difference between awaiting and not awaiting, except that we're imposing an ordering when we maybe don't need to.


I suppose awaiting means we can report `writeToScreen("SENT: ping");`, because we know `write()` didn't reject, although it's not clear whether not rejecting actually means we sent it, as https://developer.mozilla.org/en-US/docs/Web/API/WritableStreamDefaultWriter/write says:

>  Note that what "success" means is up to the underlying sink; it might indicate that the chunk has been accepted, and not necessarily that it is safely saved to its ultimate destination. 

But then if that is the reason, isn't it better to use try/catch so we can report an error?

So it could say:

```js
    try {
          await writer.write("ping");
          writeToScreen("SENT: ping");
    } catch(e) {
          writeToScreen(`Error writing to socket: ${e.message}`);
    }
```

...and then there's a use for `await`.

Ah, maybe it's not worth worrying about.

chrisdavidmills replied:
> You don't have to await to make it work: await just ensures that the promise fulfills before it executes the next line.

Ah, this is really interesting. I've been using this stuff for years and not considered that before. I've learnt something today.

I thought about it a bit and decided to just remove the unneeded `await` keywords (and the `async` keyword on the anonymous function inside the `setTimeout()` call). I don't want to add in a bunch of error handling code just cos. It doesn't add much to what I am trying to demonstrate, and increases complexity.

chrisdavidmills replied:
Update: I've actually updated the `write()` call inside the `setTimeout()` function to use the try/catch structure you suggested because sometimes you close the connection and then the app still tries to write to the stream afterwards, causing an error.

wbamberg replied:
Just to be clear: I'm not certain what the right thing to do here is, which is why I'm asking this as questions. I think async can be complex, especially when there's error-handling going on, and I don't have the time/bandwidth to say what this code ought to be doing. I just want us to think carefully about why we're making the choices that we are here.

chrisdavidmills commented on files/en-us/web/api/websockets_api/using_websocketstream/index.md:
````diff
@@ -0,0 +1,273 @@
+});
+```
+
+You can test this out by including something like the following after the previous code block:
````
Yeah, I think you're right. I've updated the demo to include a close button and added an event listener to close the connection.

chrisdavidmills commented on files/en-us/web/api/websocketstream/websocketstream/index.md:
````diff
@@ -0,0 +1,77 @@
+controller.abort();
+```
+
+> [!NOTE]
+> Alternatively, you can use the {{domxref("WebSocketStream.close()")}} method to close a connection, however this is mainly needed if you wish to specify a custom code and reason for the server to report.
````
Agreed. I've de-noted it.

chrisdavidmills commented on files/en-us/web/api/websocketstream/websocketstream/index.md:
```diff
@@ -0,0 +1,77 @@
+- `url`
+  - : A string representing the URL of the WebSocket server you want to connect to with this `WebSocketStream` instance. Allowed URL schemes are `"ws"`, `"wss"`, `"http"`, and `"https"`.
+- `options` {{optional_inline}}
+  - : An object than can contain the following properties:
```
fixed - ta!

chrisdavidmills commented on files/en-us/web/api/websocketstream/websocketstream/index.md:
````diff
@@ -0,0 +1,77 @@
+## Syntax
+
+```js-nolint
+new WebSocketStream(url, options)
````
Good call; updated.

chrisdavidmills commented on files/en-us/web/api/websocketstream/close/index.md:
```diff
@@ -0,0 +1,71 @@
+The **`close()`** method of the
+{{domxref("WebSocketStream")}} interface closes the WebSocket connection. The method optionally accepts an object specifying a custom code and reason for the server to report.
+
+> [!NOTE]
+> An alternative mechanism for closing a `WebSocketStream` is to specify an {{domxref("AbortSignal")}} in the [`signal`](/en-US/docs/Web/API/WebSocketStream/WebSocketStream#signal) option of the constructor upon creation. The associated {{domxref("AbortController")}} can then be used to close the WebSocket connection. This is generally the preferred mechanism, however `close()` can be used if you ...
```
Again, de-noted.

chrisdavidmills commented on files/en-us/web/api/websocketstream/close/index.md:
```diff
@@ -0,0 +1,71 @@
+The **`close()`** method of the
+{{domxref("WebSocketStream")}} interface closes the WebSocket connection. The method optionally accepts an object specifying a custom code and reason for the server to report.
+
+> [!NOTE]
+> An alternative mechanism for closing a `WebSocketStream` is to specify an {{domxref("AbortSignal")}} in the [`signal`](/en-US/docs/Web/API/WebSocketStream/WebSocketStream#signal) option of the constructor upon creation. The associated {{domxref("AbortController")}} can then be used to close the WebSocket connection. This is generally the preferred mechanism, however `close()` can be used if you ...
```
updated

chrisdavidmills commented on files/en-us/web/api/websocketstream/closed/index.md:
```diff
@@ -0,0 +1,88 @@
+{{APIRef("WebSockets API")}}{{non-standard_header}}
+
+The **`closed`** read-only property of the
+{{domxref("WebSocketStream")}} interface returns a {{jsxref("Promise")}} that fulfills with an object once the socket connection is closed. The object contains the closing code and reason sent by the server.
```
AFAIUI, these are always sent by the server, and then reported on the client.

wbamberg replied:
Then I am confused because in the demo we call `close()` and fetch the reason using `closed`.

chrisdavidmills replied:
Hrm, I've read some more stuff on this, and it looks like sometimes it is the client that closes the connection. The `CloseEvent.close` page lists at least one client-specific closing reason.

To hopefully mitigate the confusion, I've removed a bunch of the server-specific language from this and other pages.

chrisdavidmills committed:
Fixes for wbamberg review comments

chrisdavidmills committed:
Merge branch 'main' into websocketstream

wbamberg commented on files/en-us/web/api/websockets_api/using_websocketstream/index.md:
````diff
@@ -0,0 +1,290 @@
+writer.write("My message");
+```
+
+To read data from the socket, you can continuously call {{domxref("ReadableStreamDefaultReader.read()")}} until the stream has finished, which is indicated by `done` being `true`:
+
+```js
+while (true) {
+  const { value, done } = await reader.read();
+  if (done) {
+    break;
+  }
+
+  // Process value in some way
+}
+```
+
````
Here I think is where we should spell out the backpressure thing. Perhaps something like:

````suggestion
To read data from the socket, you can continuously call {{domxref("ReadableStreamDefaultReader.read()")}} until the stream has finished, which is indicated by `done` being `true`:

```js
while (true) {
  const { value, done } = await reader.read();
  if (done) {
    break;
  }

  // Process value in some way
}
```

Note that the client can control the rate at which it receives data by calling `read()` only when it is ready: if data is arriving faster than the client can read it, the underlying Streams API automatically exerts backpressure on the sender.

````

(assuming that is true!)

chrisdavidmills replied:
I don't think the point when `read()` is called has anything to do with backpressure. This is not automatic, but a choice made by the developer in the code?

I've read around this subject a bit, and added the following paragraph after the code snippet:

> The browser automatically controls the rate at which the client receives and sends data by applying backpressure when needed. If data is arriving faster than the client can `read()` it, the underlying Streams API exerts backpressure on the server. In addition, `write()` operations will only proceed if it is safe to do so.

wbamberg commented on files/en-us/web/api/websockets_api/using_websocketstream/index.md:
```diff
@@ -0,0 +1,290 @@
+            reason: "That's all folks",
+          });
+
+          closeBtn.disabled = false;
```
true?

chrisdavidmills commented on files/en-us/web/api/websockets_api/index.md:
```diff
@@ -2,20 +2,31 @@
 
 {{DefaultAPISidebar("WebSockets API")}}
 
 The **WebSocket API** is an advanced technology that makes it possible to open a two-way interactive communication session between the user's browser and a server. With this API, you can send messages to a server and receive event-driven responses without having to poll the server for a reply.
```
Sounds reasonable. Updated to use your wording.

chrisdavidmills committed:
Fixes for 2nd round of wbamberg review

wbamberg commented on files/en-us/web/api/websocket/index.md:
```diff
@@ -11,6 +11,9 @@ The `WebSocket` object provides the API for creating and managing a [WebSocket](
 To construct a `WebSocket`, use the [`WebSocket()`](/en-US/docs/Web/API/WebSocket/WebSocket) constructor.
 
+> [!NOTE]
+> The `WebSocket` API has no way to apply [backpressure](/en-US/docs/Web/API/Streams_API/Concepts#backpressure), therefore when messages arrive faster than the application can process them, the application will either fill up the device's memory by buffering those messages, become unresponsive due to 100% CPU usage, or both. For an alterative that provides backpressure automatically, see {{domxre ...
```
```suggestion
> The `WebSocket` API has no way to apply [backpressure](/en-US/docs/Web/API/Streams_API/Concepts#backpressure), therefore when messages arrive faster than the application can process them, the application will either fill up the device's memory by buffering those messages, become unresponsive due to 100% CPU usage, or both. For an alternative that provides backpressure automatically, see {{domxref("WebSocketStream")}}.
```

wbamberg committed:
typo

wbamberg reviewed (approved):
👍 thanks Chris!
